snuba/query/parser/expressions.py

```diff
@@ -61,44 +73,94 @@ def visit_function_name(self, node: Node, visited_children: Iterable[Any]) -> st
+            Any, Expression, Any, Union[Node, Tuple[LowPriOperator, Expression]]
+        ],
     ) -> Expression:
         _, term, _, exp = visited_children
 
-        if exp is None:
-            # A check for any None child nodes
-            # that arose from empty space ''
+        if exp == "":
             return term
-        else:
-            # return exp[1] because
-            # exp[0] is a string which
-            # represents the operator
-            return plus(term, exp[1])
+
+        for elem in exp:
+            if isinstance(elem, list):
+                # If there remains Expressions
+                # in List form that need to be
+                # put together and bubbled up
+                if elem[0] == LowPriOperator.PLUS:
+                    term = plus(term, elem[1])
+                elif elem[0] == LowPriOperator.MINUS:
+                    term = minus(term, elem[1])
+            else:
+                # This branch is hit when
+                # there is only one operator left
+                # to perform arithmetic between two
+                # Expressions that have been bubbled up
+                if exp[0] == LowPriOperator.PLUS:
+                    return plus(term, exp[1])
+                elif exp[0] == LowPriOperator.MINUS:
+                    return minus(term, exp[1])
+
+        return term
 
     def visit_high_pri_arithmetic(
         self,
         node: Node,
-        visited_children: Tuple[Any, Expression, Any, Tuple[str, Expression]],
+        visited_children: Tuple[
+            Any, Expression, Any, Union[Node, Tuple[HighPriOperator, Expression]]
+        ],
     ) -> Expression:
-        _, factor, _, term = visited_children
+        _, term, _, exp = visited_children
 
-        if term is None:
-            # A check for any None child nodes
-            # that arose from empty space ''
-            return factor
-        else:
-            # return term[1] because
-            # term[0] is a string which
-            # represents the operator
-            return multiply(factor, term[1])
+        if exp == "":
+            return term
+
+        for elem in exp:
+            if isinstance(elem, list):
+                # If there remains Expressions
+                # in List form that need to be
+                # put together and bubbled up
+                if elem[0] == HighPriOperator.MULTIPLY:
+                    term = multiply(term, elem[1])
+                elif elem[0] == HighPriOperator.DIVIDE:
+                    term = divide(term, elem[1])
+            else:
+                # This branch is hit when
+                # there is only one operator left
+                # to perform arithmetic between two
+                # Expressions that have been bubbled up
+                if exp[0] == HighPriOperator.MULTIPLY:
+                    return multiply(term, exp[1])
+                elif exp[0] == HighPriOperator.DIVIDE:
+                    return divide(term, exp[1])
+
```

Comment:
It might be a little bit clearer having the base case outside of the for loop. I think iterating through all the elements in the expression is not relevant in this scenario.